jetty-util/src/main/java/org/eclipse/jetty/util/resource/ByteBufferResource.java

```diff
@@ -0,0 +1,230 @@
+    {
+        this(rawContents.getBytes(charset));
+    }
+
+    public ByteBufferResource(byte[] bytes)
+    {
+        this(bytes, 0, bytes.length);
+    }
+
+    public ByteBufferResource(byte[] buffer, int offset, int length)
+    {
+        this(ByteBuffer.wrap(buffer, offset, length));
+    }
+
+    public ByteBufferResource(ByteBuffer byteBuffer)
+    {
+        this.content = byteBuffer;
+        this.lastModified = System.currentTimeMillis();
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("%s@%x[size=%d]", this.getClass().getSimpleName(), this.hashCode(), this.content.remaining());
+    }
+
+    @Override
+    public boolean isContainedIn(Resource r)
+    {
+        // always false
+        return false;
+    }
+
+    @Override
+    public void close()
+    {
+        // does nothing
+    }
+
+    @Override
+    public boolean exists()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isDirectory()
+    {
+        return false;
+    }
+
+    @Override
+    public long lastModified()
+    {
+        return lastModified;
+    }
+
+    @Override
+    public long length()
+    {
+        return content.remaining();
+    }
+
+    @Override
+    public URL getURL()
+    {
+        try
+        {
+            return getURI().toURL();
+        }
+        catch (MalformedURLException e)
+        {
+            throw new RuntimeException("Unrecognized URL", e);
+        }
+    }
+
+    @Override
+    public URI getURI()
+    {
+        return URI.create("buffer://");
+    }
+
+    @Override
+    public File getFile()
+    {
+        return null; // always null
+    }
+
+    @Override
+    public String getName()
+    {
+        return name;
+    }
+
+    public void setName(String name)
+    {
+        this.name = name;
+    }
+
+    @Override
+    public InputStream getInputStream()
+    {
+        return new ByteBufferInputStream(content);
+    }
+
+    @Override
+    public ReadableByteChannel getReadableByteChannel()
+    {
+        return new ByteBufferChannel(content);
+    }
+
+    @Override
+    public boolean delete() throws SecurityException
+    {
+        return true;
+    }
+
+    @Override
+    public boolean renameTo(Resource dest) throws SecurityException
+    {
+        if (dest instanceof ByteBufferResource)
+            return true;
+        return false;
+    }
+
+    @Override
+    public String[] list()
+    {
+        return new String[0]; // always empty
+    }
+
+    @Override
+    public Resource addPath(String path)
+    {
+        // memory has no paths, so it always returns the same location
+        return this;
+    }
+
+    private static class ByteBufferInputStream extends InputStream
+    {
+        private final ByteBuffer content;
+
+        public ByteBufferInputStream(ByteBuffer buffer)
+        {
+            this.content = buffer.slice();
+        }
+
+        @Override
+        public int read() throws IOException
```

Comment:
`read()`  is sufficient to function, but not very efficient.  The bulk reads should also be implemented.